File: go/two-crystal-balls/two-crystal-balls.go
package two_crystal_balls

import (
	"errors"
	"math"
)

func Two_Crystal_Balls[T comparable](breaks []T, target T) (int, error) {
	length := len(breaks)
	jump_ammount := int(math.Floor(math.Sqrt(float64(length))))

	i := jump_ammount
	for ; i < length; i += jump_ammount {
		if breaks[i] == target {
			break
		}
	}

	i -= jump_ammount

	for j := 0; j < jump_ammount && i < length; j++ {
		if breaks[i] == target {
			return i, nil
		}
		i++
	}

	return int(-1), errors.New("Target does not exist in list")
}

File: typescript/src/two-crystal-balls.ts
// vim: noexpandtab

/**
 * Find index in sorted array in O(√n) running time
 */
export function twoCrystalBalls<T = boolean>(breaks: T[], target: T): number {
	const jump_ammount = Math.floor(Math.sqrt(breaks.length));

	let i = jump_ammount;
	for (; i < breaks.length; i += jump_ammount) {
		if (breaks[i] === target) {
			break;
		}
	}

	i -= jump_ammount;

	for (let j = 0; j < jump_ammount && i < breaks.length; ++j, ++i) {
		if (breaks[i] === target) {
			return i;
		}
	}

	throw new Error('Target does not exist in list');
}
